mautic/mautic
# Send email and sms webhook events added

**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | Y
| Automated tests included? | Y
| Related user documentation PR URL | /
| Related developer documentation PR URL | /
| Issues addressed (#s or URLs) | /
| BC breaks? | N
| Deprecations? | N

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

This PR adds 2 new events to the global webhooks:

![Screenshot 2019-05-15 at 10 59 22](https://user-images.githubusercontent.com/1235442/57762840-836a5500-7700-11e9-9c9c-cee53ae906e3.png)

### Email Send Event

It will not fire a webhook for internal test emails. But it will fire it for:

- **Send email directly to a contact from the contact detail page**

Example payload:
```
{
    "mautic.email_on_send": [
        {
            "email": null,
            "contact": {
                "id": "12",
                "title": null,
                "firstname": "John",
                "lastname": "Doe",
                "email": "johndoe@gmail.com",
                [...]
            },
            "tokens": {
                "{unsubscribe_text}": "<a href=\"http://mautic-cloud.test/index_dev.php/email/unsubscribe/5cdaa50e155f9732391159\">Unsubscribe</a> to no longer receive emails from us.",
                [...]
            },
            "contentHash": "8bb9181fa1c76671352b79565e244240",
            "idHash": "5cdaa50e155f9732391159",
            "content": "<!DOCTYPE html><html>[...]</html>",
            "name": "mautic.email_on_send",
            "subject": "Test email",
            "source": [],
            "headers": {
                "List-Unsubscribe": "<http://mautic-cloud.test/index_dev.php/email/unsubscribe/5cdaa50e155f9732391159>"
            },
            "timestamp": "2019-05-14T11:22:54+00:00"
        }
    ]
}
```

- **Send email from a form action**
- **Send email from a campaign action**
- **Send segment email**

Example payload:
```
{
    "mautic.email_on_send": [
        {
            "email": {
                "isPublished": true,
                "dateAdded": "2019-05-15T10:03:20+00:00",
                "dateModified": null,
                "createdBy": 1,
                "createdByUser": "John Doe",
                "modifiedBy": null,
                "modifiedByUser": null,
                "id": 8,
                "name": "Test Email Internal Name",
                "subject": "Test Email",
                "language": "en",
                "category": null,
                "fromAddress": null,
                "fromName": null,
                "replyToAddress": null,
                "bccAddress": null,
                "utmTags": {
                    "utmSource": null,
                    [...]
                },
                "customHtml": "<!DOCTYPE html><html>[...]</html>",
                "plainText": null,
                "template": "blank",
                "emailType": "list",
                "publishUp": null,
                "publishDown": null,
                "readCount": 0,
                "sentCount": 0,
                "revision": 1,
                "assetAttachments": [],
                "variantStartDate": null,
                "variantSentCount": 0,
                "variantReadCount": 0,
                "variantParent": null,
                "variantChildren": [],
                "translationParent": null,
                "translationChildren": [],
                "unsubscribeForm": null,
                "dynamicContent": [
                    {
                        "tokenName": "Dynamic Content 1",
                        "content": "Default Dynamic Content",
                        "filters": [
                            {
                                "content": null,
                                "filters": []
                            }
                        ]
                    }
                ],
                "lists": [
                    {
                        "createdByUser": "Clare Schowalter",
                        "modifiedByUser": "Clare Schowalter",
                        "id": 1,
                        "name": "linhartescope",
                        "alias": "sdf",
                        "description": null,
                        "filters": [
                            {
                                "glue": "and",
                                "field": "email",
                                "object": "lead",
                                "type": "email",
                                "filter": "john",
                                "display": null,
                                "operator": "startsWith"
                            }
                        ],
                        "isGlobal": true,
                        "isPreferenceCenter": false
                    }
                ],
                "headers": []
            },
            "contact": {
                "id": "15",
                "date_added": "2019-04-15 12:11:46",
                "email": "linhartescope+contact3@gmail.com",
                [...]
            },
            "tokens": {
                "{dynamiccontent=\"Dynamic Content 1\"}": "Default Dynamic Content",
               [...]
            },
            "contentHash": "1692051d9cd408e413cd88c1de475edd",
            "idHash": "5cdbe478cd6fd396409369",
            "content": "<!DOCTYPE html><html>[...]</html>",
            "subject": "Test Email",
            "source": ["email", 8],
            "headers": {
                "Precedence": "Bulk",
                "List-Unsubscribe": "<https://xxx/email/unsubscribe/5cdbe478cd6fd396409369>"
            },
            "timestamp": "2019-05-15T10:05:45+00:00"
        }
    ]
}
```

### Text Send Event

Same as for emails above, just sending text messages.

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. Create a webhook that has the new events active. Use for example https://requestbin.com to track the payloads sent with this webhook.
3. Create a new segment email. Send test email. No webhook is created.
4. Send the segment email. A new webhook payload will appear for each contact in the segment.
5. Test also sending a campaign email, form action email and email from the contact detail page. All those should create webhook payloads.
6. Test also SMS send webhook by sending SMS.
